mgmt/src/processor/k8s_client.rs

```diff
@@ -1,56 +1,229 @@
+        MgmtStatusError::FetchStatusError("Failure relaying status fetch request".to_string())
+    })?;
+    let response = rx.await.map_err(|_| {
+        MgmtStatusError::FetchStatusError(
+            "Failure receiving status from config processor".to_string(),
+        )
+    })?;
+
+    match response {
+        ConfigResponse::GetDataplaneStatus(status) => Ok(*status),
+        _ => unreachable!(),
+    }
+}
+
+async fn get_current_config(
+    tx: &Sender<ConfigChannelRequest>,
+) -> Result<GwConfig, MgmtStatusError> {
+    let (req, rx) = ConfigChannelRequest::new(ConfigRequest::GetCurrentConfig);
+    tx.send(req).await.map_err(|_| {
+        MgmtStatusError::FetchStatusError("Failure relaying get config request".to_string())
+    })?;
+    let response = rx.await.map_err(|_| {
+        MgmtStatusError::FetchStatusError("Failure receiving config from processor".to_string())
+    })?;
+    match response {
+        ConfigResponse::GetCurrentConfig(opt_config) => {
+            opt_config.ok_or(MgmtStatusError::NoConfigApplied)
+        }
+        _ => unreachable!(),
+    }
+}
+
+fn to_datetime(opt_time: Option<&SystemTime>) -> chrono::DateTime<Utc> {
+    match opt_time {
+        Some(time) => chrono::DateTime::<Utc>::from(*time),
+        None => Utc.timestamp_opt(0, 0).unwrap(),
+    }
+}
+
+async fn update_gateway_status(hostname: &str, tx: &Sender<ConfigChannelRequest>) -> () {
+    let status = get_dataplane_status(tx).await;
+
+    let status = match status {
+        Ok(status) => status,
+        Err(err) => {
+            error!(
+                "Failed to fetch dataplane status, skipping status update: {}",
+                err
+            );
+            return;
+        }
+    };
+
+    let (last_applied_gen, last_applied_time) = match get_current_config(tx).await {
+        Ok(config) => (config.genid(), to_datetime(config.meta.apply_t.as_ref())),
+        Err(e) => {
+            error!("Failed to get current config, skipping status update: {e}");
+            return;
+        }
+    };
+
+    let k8s_status = match GatewayAgentStatus::try_from(&DataplaneStatusForK8sConversion {
+        last_applied_gen: Some(last_applied_gen),
+        last_applied_time: Some(&last_applied_time),
+        last_collected_time: Some(&chrono::Utc::now()),
+        status: Some(&status),
+    }) {
+        Ok(status) => status,
+        Err(err) => {
+            error!("Failed to convert status to GatewayAgentStatus: {err}");
+            return;
+        }
+    };
+
+    match replace_gateway_status(hostname, &k8s_status).await {
+        Ok(()) => (),
+        Err(err) => {
+            error!("Failed to update gateway status: {err}");
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum MgmtStatusError {
+    #[error("Failed to fetch dataplane status: {0}")]
+    FetchStatusError(String),
+    #[error("No config is currently applied")]
+    NoConfigApplied,
+}
+
+pub struct K8sClient {
+    hostname: String,
```

Comment:
part of me want to make a qualified type around `String` which describes valid hostnames.  Will leave that for another PR tho

Reply:
Yep, that is needed along with UnicastIpv[46]Net that only generates value unicast addresses.  Add it to the list :)